core/types/types.go

```diff
@@ -56,57 +50,35 @@ type BlockHeader struct {
 	StateRoot      common.Hash `json:"stateRoot"`
 	ExtrinsicsRoot common.Hash `json:"extrinsicsRoot"`
 	Digest         []byte      `json:"digest"` // any additional block info eg. logs, seal
+	hash           common.Hash
 }
 
-func (bh *BlockHeader) Hash() (common.Hash, error) {
-	enc, err := scale.Encode(bh)
-	if err != nil {
-		return [32]byte{}, err
-	}
+// SetHash sets a block header's hash. Note, you would probably want to use Hash() instead,
+// unless you know the hash has already been set
+func (bh *BlockHeader) SetHash(h common.Hash) {
+	bh.hash = h
+}
```

Comment:
Why do we need this? Seems rather unsafe

Reply:
yeah i thought the block tree needed it but upon further inspection I removed it, just had to reqrite the test a bit